### PR TITLE
resource: beefing up min/max count-based matching

### PR DESCRIPTION
**Pull request by dongahn**
This PR beefs up our support for min/max count-based Jobspec. 

- Add min/max support for our relatively new first-match policy. (Unlike other global search based match policies, our first-match policy currently stops searching when the min count request is satisfied and this keeps this policy to support min/max requests fully.)
- Add more test cases for min/max Jobspecs.

Fixes Issue #875. This also supports Issue #833 as we are pursuing ways to use min/max based jobspec to support node-exclusive scheduling.

### CODE REVIEW

File: resource/traversers/dfu_impl.cpp [hunk 1]
```diff
@@ -238,7 +238,7 @@ int dfu_impl_t::match (vtx_t u, const std::vector<Resource> &resources,
                 for (auto &c_resource : resource.with)
                     if (c_resource.type == "slot") {
                         *slot_resource = &c_resource;
-                        *nslots = c_resource.count.min;
+                        *nslots = m_match->calc_effective_max (c_resource);
```
Comment (milroy) [inline resource/traversers/dfu_impl.cpp:241]: Should `resource.count.min` be updated in `dfu_impl_t::is_enough` as well?

**Review by milroy — APPROVED** [review]
This PR LGTM.  If you have time @dongahn you can take a look at my question.  You may have a quick answer.

File: resource/traversers/dfu_impl.cpp (hunk 1 above)
Comment (dongahn) [reply resource/traversers/dfu_impl.cpp:241]: Thank you for the question. `resource.count.min` is already considered from `calc_effective_max()` method so we should be good here. If `min` == `max`, min is used. Only when `min` < `max`, max is used here.

Comment (milroy) [reply resource/traversers/dfu_impl.cpp:241]: I didn't notice that `calc_effective_max ()` was called in `dfu_impl_t::is_enough`.  I must have been looking at the wrong version of the code.

**Pull request merged**